distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala

```diff
@@ -962,6 +967,64 @@ object Lifecycle extends LifecycleCatsInstances {
     }
   }
 
+  @inline
+  private final def foldMImpl[F[_], A, B](
+    self: Lifecycle[F, A]
+  )(failure: Throwable => Lifecycle[F, B],
+    success: A => Lifecycle[F, B],
+  )(implicit F: QuasiIO[F]
+  ): Lifecycle[F, B] = {
+    import QuasiIO.syntax._
+    new Lifecycle[F, B] {
+      override type InnerResource = AtomicReference[List[() => F[Unit]]]
+
+      private[this] def bracketAppendFinalizer[a, b](finalizers: InnerResource)(lifecycle: Lifecycle[F, a])(use: lifecycle.InnerResource => F[b]): F[b] = {
```

Comment:
```suggestion
      private[this] def bracketAppendFinalizer[a, b](finalizers: InnerResource)(lifecycle: => Lifecycle[F, a])(use: lifecycle.InnerResource => F[b]): F[b] = {
```

Must suspend here to ensure that any effects in construction of lifecycle are wrapped into `bracket`'s `try` (in Identity instance). (Lifecycle/QuasiIO must work correctly with impure code)

Reply:
Hmm, this may be tricky because of path dependency

Reply:
@VladPodilnyk 
This seems to work as an encoding that supports suspension and path dependency:
```scala

      private[this] sealed trait LifecyclePair[+a] {
        val lifecycle: Lifecycle[F, a]
        val inner: lifecycle.InnerResource
      }
      private[this] def LifecyclePair[a](lifecycle0: Lifecycle[F, a])(inner0: lifecycle0.InnerResource): LifecyclePair[a] = new LifecyclePair[a] {
        override val lifecycle: lifecycle0.type = lifecycle0
        override val inner: lifecycle.InnerResource = inner0
      }

      private[this] def bracketAppendFinalizer[a, b, t](
        finalizers: InnerResource
      )(lifecycleCtor: () => Lifecycle[F, a]
      )(use: LifecyclePair[a] => F[b]
      ): F[b] = {
        F.bracket(
          acquire = {
            val lifecycle = lifecycleCtor()
            lifecycle.acquire.flatMap {
              a =>
                F.maybeSuspend {
                  // can't use `.updateAndGet` because of Scala.js
                  var oldValue = finalizers.get()
                  while (!finalizers.compareAndSet(oldValue, (() => lifecycle.release(a)) :: oldValue)) {
                    oldValue = finalizers.get()
                  }
                  LifecyclePair(lifecycle)(a)
                }
            }
          }
        )(release = _ => F.unit)(
          use = use
        )
      }
```

Reply:
Do you propose to extract it into a pattern ☝️ ?
(it's not clear for me what problem does it solves)
Also, we probably should come up with another name bc this one is not very informative, IMO.

Reply:
@neko-kai, thanks for this excellent explanation!